### PR TITLE
Teleport Join Tests ✨

### PR DESCRIPTION
**Pull request by EvolveArt**

### CODE REVIEW

File: contracts/starknet/safe_math.cairo [hunk 1]
```diff
@@ -168,3 +169,41 @@ func div{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
     let (c: Uint256, rem: Uint256) = uint256_unsigned_div_rem(a, b);
     return (c,);
 }
+
+func sub_signed256{range_check_ptr, bitwise_ptr: BitwiseBuiltin*}(lhs: Uint256, rhs: Uint256) -> (
```
Comment (maciejka) [inline contracts/starknet/safe_math.cairo]: input types should be `Int256`, right?

File: contracts/starknet/safe_math.cairo [hunk 2]
```diff
@@ -168,3 +169,41 @@ func div{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
...
+    // Check if the result fits in the correct width
+    let (res_msb: felt) = bitwise_and(res.high, 0x80000000000000000000000000000000);
+    let (res_overflow_lsb: felt) = bitwise_and(res_overflow, 1);
+    assert res_overflow_lsb * 0x80000000000000000000000000000000 = res_msb;
+
+    // Narrow and return
+    return (res=res);
+}
+
+func add_signed256{range_check_ptr, bitwise_ptr: BitwiseBuiltin*}(lhs: Uint256, rhs: Uint256) -> (
```
Comment (maciejka) [inline contracts/starknet/safe_math.cairo]: Same as above: `Int256` not `Uint256`

File: contracts/starknet/vat.cairo [hunk 3]
```diff
@@ -720,8 +719,8 @@ func frob{
     with_attr error_message("Vat/ceiling-exceeded") {
         let (debt_decreased) = _le_0(dart);
         let (ilk_debt) = mul(Art, ilk.rate);
-        let (line_ok) = le(ilk_debt, ilk.line);
-        let (Line_ok) = le(debt, ilk.line);
+        let (line_ok) = uint256_le(ilk_debt, ilk.line);
```
Comment (maciejka) [inline contracts/starknet/vat.cairo:722]: Idea behind `le` was that it is easier to read. Especially when function results will be usable in expressions, having shorter expressions will be important. Although at that point simple `<` might be supported as well.

Comment (EvolveArt) [reply contracts/starknet/vat.cairo:722]: I don't know if you've seen but there is a discussion on-going regarding felt assertions. https://github.com/OpenZeppelin/cairo-contracts/issues/433

File: contracts/starknet/teleport_join.cairo [hunk 4]
```diff
@@ -277,25 +276,100 @@ func Settle(source_domain: felt, amount: Uint256) {
...
+    // let (caller) = get_caller_address();
+    _wards.write(ward, 1);
+    Rely.emit(ward);
     _vat.write(vat_);
     _daiJoin.write(daiJoin_);
 
     VatLike.hope(vat_, daiJoin_);
     let (dai) = DaiJoinLike.dai(daiJoin_);
+    _dai.write(dai);
     TokenLike.approve(dai, daiJoin_, Uint256(2 ** 128 - 1, 2 ** 128 - 1));
```
Comment (maciejka) [inline contracts/starknet/teleport_join.cairo]: `2 ** 128 - 1` should be a constant

File: contracts/starknet/teleport_join.cairo [hunk 5]
```diff
@@ -379,13 +477,62 @@ func deny{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(user:
...
+    let (res) = mul(art, Uint256(RAY, 0));
+    return (res,);
+}
+
+// /**
+//     * @dev Internal function that registers a teleport
+//     * @param teleportGUID Struct which contains the whole teleport data
+//     * @param hashGUID Hash of the prev struct
+//     **/
+func _register{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
```
Comment (maciejka) [inline contracts/starknet/teleport_join.cairo:520]: Solidity declaration is missing

File: test/teleport_join.ts [hunk 6]
```diff
@@ -0,0 +1,1914 @@
+import { expect } from 'chai';
```
Comment (maciejka) [inline test/teleport_join.ts:1]: In general original solidity test names are missing.

**Review by maciejka — APPROVED** [review]

**Pull request merged**